crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -191,46 +228,452 @@ pub fn run_enter_schedule<S: States>(world: &mut World) {
+    let mut schedules = world.get_resource_or_insert_with(Schedules::default);
+    let mut schedule = Schedule::new(StateTransition);
+    schedule.add_systems(execute_state_transitions);
+    schedules.insert(schedule);
 }
 
 /// If a new state is queued in [`NextState<S>`], this system:
 /// - Takes the new state value from [`NextState<S>`] and updates [`State<S>`].
 /// - Sends a relevant [`StateTransitionEvent`]
 /// - Runs the [`OnExit(exited_state)`] schedule, if it exists.
 /// - Runs the [`OnTransition { from: exited_state, to: entered_state }`](OnTransition), if it exists.
+/// - Derive any dependant states through the [`ComputeDependantStates::<S>`] schedule, if it exists.
 /// - Runs the [`OnEnter(entered_state)`] schedule, if it exists.
-pub fn apply_state_transition<S: States>(world: &mut World) {
+pub fn apply_state_transition<S: FreelyMutableState>(world: &mut World) {
     // We want to take the `NextState` resource,
     // but only mark it as changed if it wasn't empty.
-    let Some(mut next_state_resource) = world.get_resource_mut::<NextState<S>>() else {
+    let Some(next_state_resource) = world.get_resource::<NextState<S>>() else {
         return;
     };
-    if let Some(entered) = next_state_resource.bypass_change_detection().0.take() {
-        next_state_resource.set_changed();
-        match world.get_resource_mut::<State<S>>() {
-            Some(mut state_resource) => {
-                if *state_resource != entered {
-                    let exited = mem::replace(&mut state_resource.0, entered.clone());
-                    world.send_event(StateTransitionEvent {
-                        before: exited.clone(),
-                        after: entered.clone(),
-                    });
-                    // Try to run the schedules if they exist.
-                    world.try_run_schedule(OnExit(exited.clone())).ok();
-                    world
-                        .try_run_schedule(OnTransition {
-                            from: exited,
-                            to: entered.clone(),
-                        })
-                        .ok();
-                    world.try_run_schedule(OnEnter(entered)).ok();
-                }
+
+    match next_state_resource {
+        NextState::Set(new_state) => {
+            let new_state = new_state.clone();
+            internal_apply_state_transition(world, Some(new_state));
+        }
+        NextState::Remove => {
+            internal_apply_state_transition::<S>(world, None);
+        }
+        _ => {
+            return;
+        }
+    }
+
+    world.insert_resource(NextState::<S>::Unchanged);
+}
+
+/// Trait defining a state that is automatically derived from other [`States`].
+///
+/// A Computed State is a state that is deterministically derived from a set of `SourceStates`.
+/// The [`StateSet`] is passed into the `compute` method whenever one of them changes, and the
+/// result becomes the state's value.
```

Comment:
IMO we should add a small end-to-end doc test here, quickly demonstrating usage for people encountering this concept through the API docs.